storage/redis/client.go

```diff
@@ -229,3 +212,82 @@ func (client *Client) allPrefixedItems(prefix, first, last storage.Key) (storage
 
 	return all, nil
 }
+
+// CompareAndSwap atomically compares and swaps oldValue with newValue
+func (client *Client) CompareAndSwap(ctx context.Context, key storage.Key, oldValue, newValue storage.Value) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	if key.IsZero() {
+		return storage.ErrEmptyKey.New("")
+	}
+
+	txf := func(tx *redis.Tx) error {
+		value, err := get(ctx, tx, key)
+		if storage.ErrKeyNotFound.Has(err) {
+			if oldValue != nil {
+				return storage.ErrKeyNotFound.New(key.String())
+			}
+
+			if newValue == nil {
+				return nil
+			}
+
+			// runs only if the watched keys remain unchanged
+			_, err = tx.Pipelined(func(pipe redis.Pipeliner) error {
+				return put(ctx, pipe, key, newValue, client.TTL)
+			})
```

Comment:
Isn't this missing a return here?

Reply:
It looks like it was intended to fall down to the `err != nil` check on line 239. Otherwise `err` would probably get a `:=` operator here.

Reply:
Yes, but if there's no error in this, then we fall even further into logic that we shouldn't hit.

Reply:
Oh dang I totally missed that. I'll add that because it's really late for Kaloyan and he might not be around atm

Reply:
Oh, this is a good catch @zeebo !